### DISCUSSION TITLE
error state

### DISCUSSION THREAD

**Discussion by martinRocks** [quartznet/quartznet#1925] [Q&A] [answered]
I have an windows service that is using the below URL as an example.  However, I have run into an exception and my trigger goes into an error state.  I know in the new version of quartz.net 3.6, there is a ResetTriggerFromErrorState method on the scheduler.  However, I don't know where to put it.

So, here are my questions.
1) Is the example that I using acceptable?
2) If so, where would the ResetTriggerFromErrorState go?
3) If the example isn't good, what do I do to make it better?
4) Is there a better way to have the trigger not die?

What I want is for the job to fail and log the error.  That is happening.  However, I need the trigger to fire again when the time is appropriate and not just die.

https://andrewlock.net/using-quartz-net-with-asp-net-core-and-worker-services/

**Comment by lahma** [comment, marked as answer]
Maybe you need to determine why this happens. One source of problems can be inability to instantiate the job, like constructor throws an error. Logging can be your friend here.

**Comment by martinRocks** [comment]
I know why my windows service is failing.  It stores data in a database and every few days the database goes into single user mode to update indexes and do a backup.  My service triggers every few minutes, so, I want it to fail and stop when there is no database.  What what I don't want is it to go on a full stuff until some person re-starts the service.

**Reply by lahma** [reply]
For transient database errors Quartz has retries for its own logic, but if it's your code that cannot handle it you could consider creating the retry policy or throwing JobExecutionException giving Quartz some guidance (`refireImmediately`), you can check inside your job whether it's a retry from context and have a sleep for example and wait for DB to come online again.

https://www.quartz-scheduler.net/documentation/quartz-3.x/how-tos/job-template.html

**Reply by martinRocks** [reply]
thank you, that was very helpful.  

For others, my database connection and all database stuff are happening in my DB layer. I made a property based on the EF method Database.CanConnect().  With that, I passed it up to my service layer.  My quartz job had access to my service, so when I can't connect to the DB, I would throw this from inside a try/catch block.

throw new JobExecutionException(new Exception("Database is down or not responding."), false);

the catch will log the error and return Task.FromException(ex);
